src/rgw/rgw_admin.cc

```diff
@@ -2639,6 +2642,234 @@ static int bucket_sync_status(rgw::sal::RGWRadosStore *store, const RGWBucketInf
+  const RGWZone& zone = store->svc()->zone->get_zone();
+
+  if (!store->ctl()->bucket->bucket_imports_data(info.bucket, null_yield)) {
+    std::cout << "Sync is disabled for bucket " << info.bucket.name << " or bucket has no sync sources" << std::endl;
+    return 0;
+  }
+
+  RGWBucketSyncPolicyHandlerRef handler;
+
+  int r = store->ctl()->bucket->get_sync_policy_handler(std::nullopt, info.bucket, &handler, null_yield);
+  if (r < 0) {
+    lderr(store->ctx()) << "ERROR: failed to get policy handler for bucket (" << info.bucket << "): r=" << r << ": " << cpp_strerror(-r) << dendl;
+    return r;
+  }
+
+  auto sources = handler->get_all_sources();
+
+  auto& zone_conn_map = store->svc()->zone->get_zone_conn_map();
+  set<rgw_zone_id> zone_ids;
+
+  if (!source_zone_id.empty()) {
+    auto z = zonegroup.zones.find(source_zone_id);
+    if (z == zonegroup.zones.end()) {
+      lderr(store->ctx()) << "Source zone not found in zonegroup "
+          << zonegroup.get_name() << dendl;
+      return -EINVAL;
+    }
+    auto c = zone_conn_map.find(source_zone_id);
+    if (c == zone_conn_map.end()) {
+      lderr(store->ctx()) << "No connection to zone " << z->second.name << dendl;
+      return -EINVAL;
+    }
+    zone_ids.insert(source_zone_id);
+  } else {
+    for (const auto& entry : zonegroup.zones) {
+      zone_ids.insert(entry.second.id);
+    }
+  }
+
+  for (auto& zone_id : zone_ids) {
+    auto z = zonegroup.zones.find(zone_id.id);
+    if (z == zonegroup.zones.end()) { /* shouldn't happen */
+      ldout(store->ctx(), 1) << "missing zone " << zone_id << dendl;
+      continue;
+    }
+    auto c = zone_conn_map.find(zone_id.id);
+    if (c == zone_conn_map.end()) { /* shouldn't happen */
+      ldout(store->ctx(), 1) << "missing zone connection " << zone_id << dendl;
+      continue;
+    }
+    if (!zone.syncs_from(z->second.name)) {
+      ldout(store->ctx(), 1) << "zone " << zone.name
+          << " does not sync from " << z->second.name << dendl;
+      continue;
+    }
+    for (auto& entry : sources) {
+      auto& pipe = entry.second;
+      if (opt_source_bucket &&
+	  pipe.source.bucket != opt_source_bucket) {
+        ldout(store->ctx(), 1) << "skipping source bucket "
+            << pipe.source.bucket << dendl;
+	continue;
+      }
+      if (pipe.source.zone.value_or(rgw_zone_id()) != z->second.id) {
+        ldout(store->ctx(), 1) << "skipping zone " << z->second.name << " != pipe source zone " << pipe.source.zone << dendl;
+        continue;
+      }
+      RGWBucketInfo source_bucket_info;
+      rgw_bucket source_bucket;
+      r = init_bucket(*pipe.source.bucket, source_bucket_info, source_bucket);
+      if (r < 0) {
+        lderr(store->ctx()) << "failed to read source bucket info: " << cpp_strerror(r) << dendl;
+        return r;
+      }
+
+      pipe.source.bucket = source_bucket;
+      pipe.dest.bucket = info.bucket;
+      r = bucket_source_sync_checkpoint(store, zone, z->second, c->second,
```

Comment:
@cbodley I'd split the checkpoint into two: first get all the remote markers of all sources, and only then wait. This way we reduce the wait on stuff that arrived when we were waiting on another source.